File: packages/http-client-java/generator/http-client-generator-test/src/test/java/com/type/model/inheritance/NestedDiscriminatorTests.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.type.model.inheritance;

import com.type.model.inheritance.nesteddiscriminator.NestedDiscriminatorClient;
import com.type.model.inheritance.nesteddiscriminator.NestedDiscriminatorClientBuilder;
import com.type.model.inheritance.nesteddiscriminator.models.Fish;
import com.type.model.inheritance.nesteddiscriminator.models.GoblinShark;
import com.type.model.inheritance.nesteddiscriminator.models.Salmon;
import com.type.model.inheritance.nesteddiscriminator.models.SawShark;
import com.type.model.inheritance.nesteddiscriminator.models.Shark;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

class NestedDiscriminatorTests {

    NestedDiscriminatorClient client = new NestedDiscriminatorClientBuilder().buildClient();

    @Test
    void getModel() {
        Fish fish = client.getModel();
        Assertions.assertEquals(1, fish.getAge());
    }

    @Test
    void putModel() {
        Shark body = new GoblinShark(1);
        client.putModel(body);
    }

    @Test
    void getRecursiveModel() {
        Salmon salmon = (Salmon) client.getRecursiveModel();
        Assertions.assertEquals(2, salmon.getFriends().size());
        Assertions.assertEquals(2, salmon.getHate().size());
        Assertions.assertEquals(SawShark.class, salmon.getPartner().getClass());
        Assertions.assertEquals(1, salmon.getAge());
        Assertions.assertEquals(2, (salmon.getPartner()).getAge());
    }


    @Test
    void putRecursiveModel() {
        Salmon salmon = new Salmon(1);
        salmon.setPartner(new SawShark(2));

        List<Fish> friends = new ArrayList<>();
        Salmon friend1 = new Salmon(2);
        friend1.setPartner(new Salmon(3));
        Map<String, Fish> friend1Hate = new LinkedHashMap<>();
        friend1Hate.put("key1", new Salmon(4));
        friend1Hate.put("key2", new GoblinShark(2));
        friend1.setHate(friend1Hate);
        friends.add(friend1);

        Shark friend2 = new GoblinShark(3);
        friends.add(friend2);
        salmon.setFriends(friends);

        Map<String, Fish> salmonHate = new LinkedHashMap<>();
        salmonHate.put("key3", new SawShark(3));
        List<Fish> salmonHateFriends = new ArrayList<>();
        salmonHateFriends.add(new Salmon(1));
        salmonHateFriends.add(new GoblinShark(4));
        salmonHate.put("key4", new Salmon(2).setFriends(salmonHateFriends));
        salmon.setHate(salmonHate);

        client.putRecursiveModel(salmon);
    }

    @Test
    void getMissingDiscriminator() {
        Fish fish = client.getMissingDiscriminator();
        Assertions.assertEquals(1, fish.getAge());
    }

    @Test
    void getWrongDiscriminator() {
        Fish fish = client.getWrongDiscriminator();
        Assertions.assertEquals(1, fish.getAge());
    }
}

File: packages/http-client-csharp/generator/TestProjects/CadlRanch.Tests/Http/_Type/Model/Inheritance/NestedDiscriminator/NestedDiscriminatorTests.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading.Tasks;
using _Type.Model.Inheritance.NestedDiscriminator;
using _Type.Model.Inheritance.NestedDiscriminator.Models;
using NUnit.Framework;

namespace TestProjects.CadlRanch.Tests.Http._Type.Model.Inheritance.NestedDiscriminator
{
    public class NestedDiscriminatorTests : CadlRanchTestBase
    {
        [CadlRanchTest]
        public Task GetMissingDiscriminator() => Test(async (host) =>
        {
            var result = await new NestedDiscriminatorClient(host, null).GetMissingDiscriminatorAsync();
            Assert.IsNotNull(result.Value);
            Assert.AreEqual(1, result.Value.Age);
        });

        [CadlRanchTest]
        public Task GetModel() => Test(async (host) =>
        {
            var result = await new NestedDiscriminatorClient(host, null).GetModelAsync();
            Assert.IsInstanceOf<GoblinShark>(result.Value);
            Assert.AreEqual(1, result.Value.Age);
        });

        [CadlRanchTest]
        public Task GetRecursiveModel() => Test(async (host) =>
        {
            var result = await new NestedDiscriminatorClient(host, null).GetRecursiveModelAsync();
            Assert.IsInstanceOf<Salmon>(result.Value);

            var salmon = (Salmon)result.Value;
            Assert.AreEqual(1, salmon.Age);
            Assert.IsTrue(salmon.Partner is Shark);

            var shark = (Shark)salmon.Partner;
            Assert.AreEqual("saw", shark.Sharktype);
            Assert.IsInstanceOf<SawShark>(shark);
            Assert.AreEqual(2, salmon.Friends.Count);
            Assert.IsInstanceOf<Salmon>(salmon.Friends[0]);

            var salmonFriend = (Salmon)salmon.Friends[0];
            Assert.AreEqual(2, salmonFriend.Age);
            Assert.AreEqual(2, salmonFriend.Hate.Count);
            Assert.IsTrue(salmon.Hate.ContainsKey("key3"));
            Assert.IsInstanceOf<SawShark>(salmon.Hate["key3"]);
            Assert.IsTrue(salmon.Hate.ContainsKey("key4"));
            Assert.IsInstanceOf<Salmon>(salmon.Hate["key4"]);
        });

        [CadlRanchTest]
        public Task GetWrongDiscriminator() => Test(async (host) =>
        {
            var result = await new NestedDiscriminatorClient(host, null).GetWrongDiscriminatorAsync();
            Assert.IsInstanceOf<UnknownFish>(result.Value);
            Assert.AreEqual(1, result.Value.Age);
            Assert.AreEqual("wrongKind", result.Value.Kind);
        });

        [CadlRanchTest]
        public Task PutModel() => Test(async (host) =>
        {
            var body = new GoblinShark(1);
            var response = await new NestedDiscriminatorClient(host, null).PutModelAsync(body);
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task PutRecursiveModel() => Test(async (host) =>
        {
            var body = new Salmon(1)
            {
                Partner = new SawShark(2),
                Friends =
                {
                    new Salmon(2)
                    {
                        Partner = new Salmon(3),
                        Hate = { { "key1", new Salmon(4) }, { "key2", new GoblinShark(2) } }
                    },
                    new GoblinShark(3)
                },
                Hate =
                {
                    { "key3", new SawShark(3) },
                    { "key4", new Salmon(2) { Friends = { new Salmon(1), new GoblinShark(4) } } }
                }
            };
            var response = await new NestedDiscriminatorClient(host, null).PutRecursiveModelAsync(body);
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });
    }
}
